staging/src/k8s.io/client-go/discovery/cached/disk/compactable_disk_cache.go

```diff
@@ -0,0 +1,234 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package disk
+
+import (
+	"os"
+	"path/filepath"
+	"sort"
+	"time"
+
+	"golang.org/x/sync/singleflight"
+
+	"github.com/gregjones/httpcache"
+	"github.com/peterbourgon/diskv"
+	"k8s.io/klog/v2"
+)
+
+const (
+	oneDay                 = 24 * time.Hour
+	lastCompactionFilename = "lastcompactiontime"
+)
+
+// CompactableDiskCache implements the httpcache.Cache interface. It delegates to
+// sumDiskCache for Cache functions (Get, Set, Delete), and occasionally removes
+// old cached files. The specified "cacheSize" is NOT a hard limit; the disk
+// cache can breach this threshold. When cache compaction occurs, it will resize
+// the disk cache below this "cacheSize".
+//
+// Trigger:
+//
+//	disk cache compactions are triggered when a file is accessed and the
+//	last compaction occurred over a day ago. Every cache file access checks
+//	when the last compaction occurred by checking the modification time
+//	on an empty file in the cache -- "lastcompactiontime". File access
+//	time for a cached file is not used directly because it is not well
+//	supported by all OSes.
+//
+// Concurrency:
+//
+//	Within the same process: the goroutines of a single process are
+//	coordinated using "sync/singleflight" functionality.
+//
+//	Among different processes: if (for example) a separate kubectl command
+//	compacts the disk cache at the same time, separate processes can compete
+//	to delete the same files. This code must be tolerant to attempting to
+//	delete files already deleted.
+type CompactableDiskCache struct {
+	cacheDir  string
+	cacheSize int64
+	delegate  httpcache.Cache
+}
+
+// Get returns the []byte representation of a cached response and a bool
+// set to true if the value isn't empty. This function can trigger a
+// disk cache compaction which runs in a separate goroutine. Touches
+// cached file to update file modification to match access time.
+func (cdc *CompactableDiskCache) Get(key string) ([]byte, bool) {
+	// If the file exists, update file modification time = access time.
+	_ = touchFile(sanitize(key))
+	go possiblyCompactDiskCache(cdc.cacheDir, cdc.cacheSize)
```

Comment:
What prevents the process from exiting before this completes, if it's e.g. a short kubectl command?

Reply:
Nothing prevents it now. But adding a `sync.WaitGroup` here within the scope of the round tripper could cause unacceptable wait times during disk compactions. This issue may be insuperable for the current approach. We may not be able to clean up cache files in the same place we are creating (and accessing and deleting) them.

/hold

Reply:
I don't know if we should keep kubectl alive just for that. I think as long as it does some clean-up, it can resume the next time it runs?

Reply:
The worry is that the cache compaction would be in the middle of a file delete and it could corrupt the cache file when the compaction goroutine exits.

Reply:
Deleting a file is an OS call, I don't think it's going to stop halfway through and corrupt the filesystem.

I am mainly worried about the case where someone never runs a slow command and the cache is big enough that it doesn't finish scanning it before the process exits.

I guess my first thought about how this conundrum could be solved is to have a `kubectl compact-cache` command and when we think it's needed, we fork and run a background process executing that command. I don't know if this is needed though.